File: python/lowest-common-ancestor-of-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        returnNode = None

        def recursiveAncestor(root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> Optional[bool]:
            nonlocal returnNode
            if not root:
                return False
            leftCheck = recursiveAncestor(root.left, p, q)
            rightCheck = recursiveAncestor(root.right, p, q)

            if leftCheck is None or rightCheck is None:
                return None

            elif leftCheck == 1 and rightCheck == 1:
                returnNode = root
            elif root.val == p.val or root.val == q.val:
                if leftCheck or rightCheck:
                    returnNode = root
                else:
                    return True
            else:
                return leftCheck or rightCheck

        recursiveAncestor(root, p, q)
        if not returnNode:
            return root
        return returnNode

File: c++/lowest-common-ancestor-of-a-binary-tree.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if (root == nullptr) return root;

        if (root == p || root == q) return root;

        root->left = lowestCommonAncestor(root->left, p, q);
        root->right = lowestCommonAncestor(root->right, p, q);

        if (root->left != nullptr && root->right != nullptr) return root;
        return root->right == nullptr ? root->left : root->right;
    }
};